### PR TITLE
Bug 1835004: Add keepYoungerThanDuration and deprecate keepYoungerThan

### PR DESCRIPTION
**Pull request by dmage**
time.Duration is encoded as an integer (nanoseconds). This is not a very convenient way to define durations.

The PR deprecated the existing field keepYoungerThan and adds keepYoungerThanDuration, that should work as it was initially designed and documented.

### CODE REVIEW

**Comment by dmage** [comment]
```
Error: unable to parse option "output:dir=/tmp/tmp.019kvpehMx": [illegal octal number (at <input>:1:10)]
```

/retest

**Comment by dmage** [comment]
/assign @deads2k 
I added a linter that should protect us from using weird types, please take a look.

/assign @adambkaplan 
/cc @ricardomaraschini

File: imageregistry/v1/types_imagepruner.go [hunk 1]
```diff
@@ -44,13 +44,17 @@ type ImagePrunerSpec struct {
...
 	Suspend *bool `json:"suspend,omitempty" protobuf:"bytes,2,opt,name=suspend"`
 	// keepTagRevisions specifies the number of image revisions for a tag in an image stream that will be preserved.
-	// Defaults to 5.
+	// Defaults to 3.
 	// +optional
 	KeepTagRevisions *int `json:"keepTagRevisions,omitempty" protobuf:"bytes,3,opt,name=keepTagRevisions"`
-	// keepYoungerThan specifies the minimum age of an image and its referrers for it to be considered a candidate for pruning.
-	// Defaults to 96h (96 hours).
+	// keepYoungerThan specifies the minimum age in nanoseconds of an image and its referrers for it to be considered a candidate for pruning.
+	// DEPRECATED: This field is deprecated in favor of keepYoungerThanDuration. If both are set, this field is ignored and keepYoungerThanDuration takes precedence.
```
Comment (adambkaplan) [inline imageregistry/v1/types_imagepruner.go:51]: @dmage I think you'll want to vendor this PR into `cluster-image-registry-operator` and demonstrate how we'd migrate users from the old field to the new field.

Comment (dmage) [reply imageregistry/v1/types_imagepruner.go:51]: As keepYoungerThan is part of a stable API, we cannot remove it and should support it. Until we want to remove it, I don't see any reason to migrate.

File: hack/typelinter/typelinter.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+					valid = false
+				}
+			}
+		}
+	}
+	return valid
+}
+
+func main() {
+	flag.Var(&cursedFields, "cursed", "exclude the field from being checked (e.g. -cursed=github.com/openshift/api/image/dockerpre012.ImagePre012:Created), can be used multiple times")
```
Comment (soltysh) [inline hack/typelinter/typelinter.go:77]: excluded

Comment (dmage) [reply hack/typelinter/typelinter.go:77]: it's the imperative mood: `exclude the field ...`
would you prefer `the field excluded from being checked`?

Comment (dmage) [reply hack/typelinter/typelinter.go]: updated

File: hack/verify-types.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh -eu
+
+builtins='[a-z0-9]+|struct{}'
+pkgs='k8s\.io/api/.*|k8s\.io/apimachinery/.*|github\.com/openshift/api/.*'
+
+go run ./hack/typelinter \
```
Comment (deads2k) [inline hack/verify-types.sh:6]: comment saying what we're trying to stop from happening

File: hack/typelinter/typelinter.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+var (
+	excludedFields stringListVar
+	typesWhitelist string
+
+	typesWhitelistRegexp *regexp.Regexp
+)
+
+func validStruct(structName string, s *types.Struct) bool {
```
Comment (deads2k) [inline hack/typelinter/typelinter.go:38]: doc an example of what the structName looks like for `string`, `PodSpec`, and `time.Duration`

File: imageregistry/v1/types_imagepruner.go (hunk 1 above)
Comment (deads2k) [reply imageregistry/v1/types_imagepruner.go:51]: > As keepYoungerThan is part of a stable API, we cannot remove it and should support it. Until we want to remove it, I don't see any reason to migrate.

True.  However, if @adambkaplan wants to see how the priority handling works in detail, I think that's fairly reasonable in this case.

**Comment by deads2k** [comment]
Thanks for the linter.

The doc looks good. The type looks ok to me.

/approve

I'll leave lgtm to @adambkaplan.  I would suggest deferring to his request to see this field in usage.

**Comment by dmage** [comment]
@adambkaplan comments are added, the use is demonstrated in https://github.com/openshift/cluster-image-registry-operator/pull/560

**Comment by adambkaplan** [comment]
@bmcelvee note that for 4.5 we will need to update the image pruner docs to use this field.

**Pull request merged**